server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -161,10 +166,18 @@ protected void onItemFailure(DiscoveryNode discoveryNode, Exception e) {
 
             @Override
             protected CheckedConsumer<ActionListener<NodesResponse>, Exception> onCompletion() {
-                // ref releases all happen-before here so no need to be synchronized
                 return l -> {
-                    try (var ignored = Releasables.wrap(Iterators.map(responses.iterator(), r -> r::decRef))) {
-                        newResponseAsync(task, request, actionContext, responses, exceptions, l);
+                    if (responsesHandled.compareAndSet(false, true)) {
+                        // ref releases all happen-before here so no need to be synchronized
+                        try (var ignored = Releasables.wrap(Iterators.map(responses.iterator(), r -> r::decRef))) {
+                            newResponseAsync(task, request, actionContext, responses, exceptions, l);
+                        }
+                    } else {
+                        logger.debug("task cancelled after all responses were collected");
+                        assert task instanceof CancellableTask : "expect CancellableTask, but got: " + task;
+                        final var cancellableTask = (CancellableTask) task;
+                        assert cancellableTask.isCancelled();
+                        throw new TaskCancelledException("task cancelled [" + cancellableTask.getReasonCancelled() + "]");
```

Comment:
This change is to address the edge case commented [here](https://github.com/elastic/elasticsearch/pull/130303#discussion_r2174359775). But I struggle to write a test for it. Essentially we need the cancel to comes in after all node responses are collected but before the `AtomicBoolean responsesHandled` is checked. One option is to extract the creation of `CancellableFanOut` into its own protected method plus wrapping the returned value with a delgating `CancellableFanOut`. But this requires making the 4 protected methods in `CancellableFanOut` package private. I am a bit suspicous on whether this is the right path to go down. I am open to suggestions.

Reply:
I'd be content with a test which concurrently completes the action and cancels it, and asserts that we always either get an exception or we get a successful response. I expect such a test would find the bug here pretty reliably.